targets/psa_switch/psa_switch.cpp

```diff
@@ -308,16 +308,22 @@ PsaSwitch::ingress_thread() {
     input_buffer.pop_back(&packet);
     if (packet == nullptr) break;
 
-    Parser *parser = this->get_parser("ingress_parser");
-    Pipeline *ingress_mau = this->get_pipeline("ingress");
-    phv = packet->get_phv();
-
     port_t ingress_port = packet->get_ingress_port();
     (void) ingress_port;
     BMLOG_DEBUG_PKT(*packet, "Processing packet received on port {}",
                     ingress_port);
 
+    phv = packet->get_phv();
+
+    Parser *parser = this->get_parser("ingress_parser");
     parser->parse(packet.get());
+
+    if (phv->has_field("psa_ingress_output_metadata.drop")) {
+      phv->get_field("psa_ingress_output_metadata.drop")
+        .set(true);
+    }
+
+    Pipeline *ingress_mau = this->get_pipeline("ingress");
     ingress_mau->apply(packet.get());
```

Comment:
I believe the reset_exit() call below should be done just after ingress_mau->apply().  As the code is now, that is never being done for multicast packets.

Also the deparser should be done before the multicast packet handling.  Right now that is never being done for multicast packets, either.

Reply:
Yes, thanks for the catch... didn't even realize we had allowed mcast packets to skip deparsing. Will fix this.

Reply:
@jafingerhut does this mean `packet->reset_exit()` also needs to be called after the egress pipeline in the event it is recirculated? I saw a comment in packet.cpp that says

> Do not forget to call reset_exit() at the end of the pipeline, if you intend the packet to go through another pipeline
 afterwards; otherwise this other pipeline will just be skipped as well.

We are somehow passing recirc tests even without having done this in egress. I'll investigate into what it does - but am also wondering if you have insight on whether it's needed after `egress_mau->apply()`?

Reply:
I was wondering the same thing last night myself while reviewing what reset_exit did and looking at the corresponding code for simple_switch, which also does not have a reset_exit call after egress_mau processing.

@antoninbas Do you know whether that could cause a bug for recirculated packets that make an exit call in egress control processing before recirculating?  Or is the fact that recirculated packets are implemented by copying the packet data structure somehow avoiding a bug there?

It certainly seems cleaner to do reset_exit() just after egress control processing, just in case.

Reply:
Not that I think that forgetting to do reset_exit() will only cause a problem if the user's program invoked an 'exit' statement in their program, and then the packet was recirculated, and there is probably no existing test case that covers this.  Perhaps there should be....

Reply:
**recirculate is now implemented without a packet copy** (for psa switch), as it should be, so you definitely need to call `reset_exit`
if you write the test you are describing, I expect it to fail, the ingress pipe will be skipped by the recirculated packet

Reply:
addressed all comments/test suggestions in my most recent push.